### ISSUE TITLE
127.0.0.1:0 causes exception

### ISSUE THREAD

**Issue by stevenvachon** [request/request#1596]
```js
require("request")("http://127.0.0.1:0");
```
```
events.js:141
      throw er; // Unhandled 'error' event
            ^
Error: connect EADDRNOTAVAIL 127.0.0.1 - Local (0.0.0.0:54620)
    at Object.exports._errnoException (util.js:843:11)
    at exports._exceptionWithHostPort (util.js:866:20)
    at connect (net.js:824:14)
    at lookupAndConnect (net.js:917:5)
    at Socket.connect (net.js:886:5)
    at Agent.exports.connect.exports.createConnection (net.js:61:35)
    at Agent.createSocket (_http_agent.js:177:16)
    at Agent.addRequest (_http_agent.js:147:23)
    at new ClientRequest (_http_client.js:132:16)
    at Object.exports.request (http.js:30:10)
```

**Comment by julien-f** [comment]
What does it mean to connect to the port 0?

0 can be used to bind to a random available port but it does not make sense to use it to connect…

**Comment by stevenvachon** [comment]
Makes perfect sense in a test suite

**Comment by mmalecki** [comment]
@stevenvachon would you mind explaining your use-case for connecting to port 0 in a test suite?

This one appears to be a difference in behavior between Node.js and Io.js - node v0.10.32 returns a `ECONNREFUSED` error to the callback while io.js v1.4.3 throws a `RangeError`. I believe that Io.js' behavior is more accurate as trying to connect to 0 *does* sound like a programmer's error.

**Comment by stevenvachon** [comment]
Oops, I misunderstood the question. *Serving* on port 0 is perfect for test suites, however *connecting* (as was said) to port 0 is stupid. I forget now *why* I'd ran that line of code, but it's at least evident at this point that the result is inconsistent than that of previous node.js versions. However, it's an io.js issue.

If anyone is curious, serving on port 0 is a standard way of connecting to any port that the OS has available, often a high number. This is perfect for a test suite as 80 or 3000 are often not available and sniffing for such is faulty.
```js
var server = http.createServer();
server.listen(0, host, function() {
	port = server.address().port;
});
```

**Comment by stevenvachon** [comment]
Clarifying here as well: the issue here is that this error is thrown, while all other errors are asynchronously passed to the callback function.

**Comment by simov** [comment]
Fixed in v2.72 https://github.com/request/request/pull/2164
